populator-machinery/controller.go

```diff
@@ -714,37 +698,33 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
 		// TODO: Handle PVC' update while the original PVC changed
 		// If the PVC is unbound and PVC' doesn't exist yet, create PVC'
 		if "" == pvc.Spec.VolumeName && pvcPrime == nil {
-			pvcPrime = &corev1.PersistentVolumeClaim{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      pvcPrimeName,
-					Namespace: c.populatorNamespace,
-				},
-				Spec: corev1.PersistentVolumeClaimSpec{
-					AccessModes:      pvc.Spec.AccessModes,
-					Resources:        pvc.Spec.Resources,
-					StorageClassName: pvc.Spec.StorageClassName,
-					VolumeMode:       pvc.Spec.VolumeMode,
-				},
-			}
-			if waitForFirstConsumer {
-				pvcPrime.Annotations = map[string]string{
-					annSelectedNode: nodeName,
-				}
-			}
-			if c.mutatorConfig != nil && c.mutatorConfig.PvcPrimeMutator != nil {
-				mp := PvcPrimeMutatorParams{
-					PvcPrime:     pvcPrime,
-					StorageClass: storageClass,
-				}
-				pvcPrime, err = c.mutatorConfig.PvcPrimeMutator(mp)
-				if err != nil {
-					c.recorder.Eventf(pvc, corev1.EventTypeWarning, reasonPVCPrimeMutatorError, "Failed to mutate populator pvcPrime: %s", err)
+			if c.providerFunctionConfig != nil && c.providerFunctionConfig.CreatePVCPrimeFn != nil {
+				if pvcPrime, err = c.providerFunctionConfig.CreatePVCPrimeFn(ctx, *params); err != nil {
```

Comment:
The only thing I'm not sure about is that there is a check of `"" == pvcPrime.Spec.VolumeName`[1].  Previous implementation was based on dynamic provisioning PV which will guarantee the check only get passed when the PVC' gets created. With `CreatePVCPrimeFn` if there is a static provisioned PV, is it possible that `pvcPrime.Spec.VolumeName` gets set, kubeClient.create.pvc' gets stuck for some reason, but the   `"" == pvcPrime.Spec.VolumeName` pass?

[1]https://github.com/kubernetes-csi/lib-volume-populator/blob/master/populator-machinery/controller.go#L773

Reply:
That would be up to the client implementation. But, if the client uses static provisioning, the proper way to do this is to create a PV with a `claimRef`, and then a PVC without a `volumeName` set. The volume controller then binds the two, and sets `volumeName`.

Reply:
It's a recommendation but not a requirement that users leave PVC.spec.volumeName unset.

Reply:
This flow probably should get added to the KEP or README later.